File: src/ucm/erikkarl/Either.java
package ucm.erikkarl;


public abstract class Either<A, B> {

    private Either() {
    }

    @Override
    public abstract String toString();

    public abstract Object getValue();

    public static final class Left<A, B> extends Either<A, B> {
        private final A value;

        public Left(A value) {
            this.value = value;
        }

        @Override
        public A getValue() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(this.value);
        }
    }


    public static final class Right<A, B> extends Either<A, B> {
        private final B value;

        public Right(B value) {
            this.value = value;
        }

        @Override
        public B getValue() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(this.value);
        }
    }
}

File: src/ucm/erikkarl/Either.kt
package ucm.erikkarl

sealed class Either<A, B> {
    class Left<A, B>(val left: A) : Either<A, B>() {
        override fun toString(): String {
            return left.toString()
        }
    }

    class Right<A, B>(val right: B) : Either<A, B>() {
        override fun toString(): String {
            return right.toString()
        }
    }

    abstract override fun toString(): String
}